src/types.mligo

```diff
@@ -139,6 +139,25 @@ type delegate =
   }
 type delegates = (delegate, unit) big_map
 
+
+// Proposal Doubly Linked List
+
+type first_proposal_node = (proposal_key * (proposal_key option)) // The first key and its next key
+type last_proposal_node = (proposal_key * proposal_key) // The last key and its prev key
+
+type proposal_doubly_linked_list =
+  [@layout:comb]
+  { first: first_proposal_node option
+  ; last: last_proposal_node option
+  ; map: (proposal_key, (proposal_key * proposal_key)) big_map
+  }
```

Comment:
I've considered this quite a bit and I think that I came up with a different representation that would be more efficient in size as well as cost (of running functions on it).

So AFAIK this data structure needs to support a handful of operations:
- `mem` to find if a `proposal_key` is in the dlist
- `pop` to get the first item and remove it from the dlist
- `push` to put a new item at the end of the dlist
- `delete` to remove an item regardless of where it is in the dlist

My main considerations are this:
- `push` and `pop` only need to ever update a "link" in one direction (each)
- `remove` needs most of the times to update a "link" in both direction (except when there is a single element)
- `remove` is less commonly used than `push` and `pop`, so the last two should be favored
- `mem` however needs to be as cheap as possible
- when the list is empty, we don't need to keep track of any information (except that it's empty)
- it would be cheaper to have no `option`s in the types as much as possible

So, here's what came to my mind to store this instead:
1. make `proposal_doubly_linked_list` only track non-empty dlists, so in the `storage` the type of the field would not be `proposal_doubly_linked_list` but rather `proposal_doubly_linked_list option`. So when the list is empty this is `None`.
2. make `first` and `last` be only a `proposal_key`. If there is a single element their value is equal. 
3. in the `big_map` store "link" separately depending on direction.
   Let's use a type synonym `direction` for a `bool` and say that "next" is `true` and "previous" is false.
   What I propose is that the key for the `big_map` is `(proposal_key, direction)` and that the value is a `proposal_key`.
   - If a proposal is the `first` then there is no `(proposal_key, false)` in the map for it.
   - If a proposal is the `last` then there is no `(proposal_key, true)` in the map for it.
   - If there is a single proposal in the whole dlist, then there is no value for it in the list at all (combining the two previous points)
   - In all other cases there are two entries for each `proposal_key`, one with `true` and another with `false`.

This is perhaps a bit more complex, but as you can see:
- there are no `option` in the types (except a single one when the list is empty)
- I think that the cost of the operations above will be much smaller (see next comment)

Reply:
To summarize the operations and how they are cheaper (here I'm using `proposal_id` as the value of the input `proposal_key`)

for `mem`:
- if there is no dlist -> false
- if the `proposal_id` is equal to `first` -> true
- if there is a `(proposal_id, false)` in `big_map` -> true
- otherwise -> false

for `push`:
- if there is no dlist -> `first` and `last` are `proposal_id` and the `big_map` is empty
- otherwise:
   -  push `(last, false)` in the `big_map` with a value of `Some proposal_id`
   - set `last` to `proposal_id`

Similarly `pop` only needs to care about `first` and the `(first, true)` key of the `big_map`.

Only `delete` may need to update two values of the `big_map`, but it will still take less comparisons.


@rinn7e  what do you think? Does this sound like an improvement to you?

Reply:
@pasqu4le Spent sometimes thinking about your suggestion, and definitely sounds good to me. I was thinking about how to store the proposal link separately as well initially, but didn't come up with a good solution. 

I'll start working on the changes, thanks a lot Pasquale